### DISCUSSION TITLE
Question on relay cursor pagination

### DISCUSSION THREAD

**Discussion by hyochan** [prisma/prisma#3190] [Q&A] [answered]
I came across the [Prisma document on relay pagination](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/pagination#cursor-based-pagination). This pretty works well except one concern.

If I want to get user connections as below,
```ts
export const usersQueryField = queryField((t) => {
  t.connectionField('users', {
    type: 'User',
    async resolve(_, args, ctx) {
      const users = await ctx.prisma.user.findMany();
      return connectionFromArray(
        users,
        args,
      );
    },
  });
});
```

`Prisma` will always find many `users` and not the filtered one. Is the above the right implementation?
I am concerned if I have to put `whereArgs` when finding many users and how if I should. I've just found that the `cursor` values provided by [graphql-relay](https://github.com/graphql/graphql-relay-js) is newly provided and are not from the fields in our user table.

**Comment by ryands17** [comment, marked as answer]
Hey @hyochan 👋 
I think you would need to add the where args to the `findMany` user operation as that will be how Prisma can identify how you are paginating. If you just have this `ctx.prisma.user.findMany()` it will just give you values without any operation, equivalent to a `select *` without filters

**Reply by hyochan** [reply]
Thanks for the answer. However, I think I didn't make myself clear enough.

I am more concerned on whether `prisma` is getting all the users in the table at once and not the required number of users. Say I want to have the first page of 10 users, but it looks like `Prisma` is getting all users first hand. My actual question was if this approach is fine or if I am miss-understanding something.

**Reply by ryands17** [reply]
Yes Prisma will get all the values as you haven't specified any filters in `findMany`.

This snippet:

```ts
const users = await ctx.prisma.user.findMany();
```

Will fetch all the values from the DB at once.

Where as this will fetch the first 10 items:

```ts
const users = await ctx.prisma.user.findMany({
  take: 10,
});
```

Then you pass the last item's id in the cursor to get the next 10 items:

```ts
const otherUsers = await ctx.prisma.user.findMany({
  take: 10,
  skip: 1,
  cursor: {
    id: users[users.length - 1].id,
  },
});
```

Does this make it more clear?

**Reply by hyochan** [reply]
Thanks now I understood that the query should be fixed in terms of performance.
However, still not sure how to beautifully attach `connectionFromArray` in `graphql-relay` to `prisma`.

My query looks something like below.

```ts
export const usersQueryField = queryField((t) => {
  t.connectionField('users', {
    type: 'User',
    additionalArgs: {
      email: stringArg(),
      name: stringArg(),
    },
    async resolve(_, args, ctx) {
      const { first, after, before, last, email, name } = args;
      const users = await ctx.prisma.user.findMany({
        take: first,
        cursor: ?, // The relay connection generates unique cursor by itself.
        where: {
          email: {
            contains: email,
          },
          name: {
            contains: name,
          },
          deletedAt: null,
        },
      });
      return connectionFromArray(
        users,
        args,
      );
    },
  });
});
```

I'll try investigate this. Just unsure where I can see additional doc related to this.

**Reply by Weakky** [reply]
The `PageInfo` type from `connectionFromArray` is unfortunately incompatible with the Nexus typings because of some nulalbility differences. You can however get it to work without graphql-relay using the `nodes` property of `connectionField`:

```ts
export const usersQueryField = queryField((t) => {
  t.connectionField('users', {
    type: 'User',
    additionalArgs: {
      email: stringArg(),
      name: stringArg(),
    },
    cursorFromNode(node) { // Can be configured globally in the plugin constructor
      return node.id
    },
    async nodes(_, args, ctx) {
      const { first, after, before, last, email, name } = args;
      const users = await ctx.prisma.user.findMany({
        take: first + 1, // Important to always fetch one more item so that we can know if there's a next page
        cursor: args.after.id,
        where: {
          email: {
            contains: email,
          },
          name: {
            contains: name,
          },
          deletedAt: null,
        },
      });
      return users
    },
  });
});
```

You can find an example here: https://github.com/prisma-labs/nextjs-graphql-api-examples/tree/master/packages/graphql-with-prisma-relay

Alternatively, if you don't want to do the mapping between the relay pagination args and the prisma's ones, you can try https://github.com/devoxa/prisma-relay-cursor-connection, which you can use in the `resolve` property just like you were with the `connectionFromArray` method

**Reply by hyochan** [reply]
Hey this is awesome 🙌  Thanks a lot! I've ended up copying [util.ts](https://github.com/prisma-labs/nextjs-graphql-api-examples/blob/master/packages/graphql-with-prisma-relay/services/utils.ts). I hope this implementation to go somewhere inside `nexus` though.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
